extension/opampextension/opamp_agent.go

```diff
@@ -0,0 +1,205 @@
+				return o.composeEffectiveConfig(), nil
+			},
+			OnMessageFunc: o.onMessage,
+		},
+		// Include ReportsEffectiveConfig once the extension has access to the
+		// collector's effective configuration.
+		Capabilities: protobufs.AgentCapabilities_AgentCapabilities_Unspecified,
+	}
+
+	if err := o.createAgentDescription(); err != nil {
+		return err
+	}
+
+	if err := o.opampClient.SetAgentDescription(o.agentDescription); err != nil {
+		return err
+	}
+
+	o.logger.Debug("Starting OpAMP client...")
+
+	if err := o.opampClient.Start(context.Background(), settings); err != nil {
+		return err
+	}
+
+	o.logger.Debug("OpAMP client started")
+
+	return nil
+}
+
+func (o *opampAgent) Shutdown(ctx context.Context) error {
+	o.logger.Debug("OpAMP agent shutting down...")
+	if o.opampClient != nil {
+		o.logger.Debug("Stopping OpAMP client...")
+		err := o.opampClient.Stop(ctx)
+		return err
+	}
+	return nil
+}
+
+func newOpampAgent(cfg *Config, logger *zap.Logger) (*opampAgent, error) {
+	uid := ulid.Make() // TODO: Replace with https://github.com/open-telemetry/opentelemetry-collector/issues/6599
+
+	if cfg.InstanceUID != "" {
+		puid, err := ulid.Parse(cfg.InstanceUID)
+		if err != nil {
+			return nil, err
+		}
+		uid = puid
+	}
+
+	agent := &opampAgent{
+		cfg:             cfg,
+		logger:          logger,
+		agentType:       "io.opentelemetry.collector",
+		agentVersion:    "1.0.0", // TODO: Replace with actual collector version info.
+		instanceId:      uid,
+		effectiveConfig: localConfig, // TODO: Replace with https://github.com/open-telemetry/opentelemetry-collector/issues/6596
+	}
+
+	return agent, nil
+}
+
+func stringKeyValue(key, value string) *protobufs.KeyValue {
+	return &protobufs.KeyValue{
+		Key: key,
+		Value: &protobufs.AnyValue{
+			Value: &protobufs.AnyValue_StringValue{StringValue: value},
+		},
+	}
+}
+
+func (o *opampAgent) createAgentDescription() error {
+	hostname, err := os.Hostname()
+	if err != nil {
+		return err
+	}
+
+	ident := []*protobufs.KeyValue{
+		stringKeyValue("service.instance.id", o.instanceId.String()),
+		stringKeyValue("service.name", o.agentType),
+		stringKeyValue("service.version", o.agentVersion),
+	}
+
+	nonIdent := []*protobufs.KeyValue{
+		stringKeyValue("os.arch", runtime.GOARCH),
+		stringKeyValue("os.family", runtime.GOOS),
+		stringKeyValue("host.name", hostname),
+	}
+
+	o.agentDescription = &protobufs.AgentDescription{
+		IdentifyingAttributes:    ident,
+		NonIdentifyingAttributes: nonIdent,
+	}
+
+	return nil
+}
+
+func (o *opampAgent) updateAgentIdentity(instanceId ulid.ULID) {
+	o.logger.Debug("OpAMP agent identity is being changed",
+		zap.String("old_id", o.instanceId.String()),
+		zap.String("new_id", instanceId.String()))
+	o.instanceId = instanceId
+}
+
+func (o *opampAgent) composeEffectiveConfig() *protobufs.EffectiveConfig {
+	return &protobufs.EffectiveConfig{
+		ConfigMap: &protobufs.AgentConfigMap{
```

Comment:
How does this work with secrets and values injected from other sources? like `${file:blah}` ? 

Are you expecting to the resolved configuration to be sent or to the unparsed configuration file?

Reply:
The effective config here will be the Collector's full running config and will have these values resolved.

Reply:
Do you have a reference on how this configuration is meant to be used?

I mostly don't want the effective config with all the potential secrets to be exported to a remote server which could potentially then accidentally leak those secrets. The same for the inverse.

Reply:
Sure, the specification section on configuration is [here](https://github.com/open-telemetry/opamp-spec/blob/main/specification.md#configuration). The gist as I understand it is that it is intended for the user to see what the Collector is running as a debugging mechanism. Your concern around secrets is valid, I'm not entirely sure how to address it. Currently the process for getting the effective configuration to the extension has no semantic understanding of the configuration.

Reply:
The dumb approach to achieving this would be to have the extension hash all string/number values in the configuration, then on the server replace the hashes with the values it has when displaying the config. The operator would have to confirm the remaining values using their own copy of the other config sources. I think we could guarantee that no secrets have been leaked to the server with this approach.

Reply:
There is a type named `configopaque.String` which is mask the values if you're using the `TextMarshaller` interface however, I worry this might not be enough.

To help get this PR out of draft, I am happy to go ahead provided there is an issue following up the potential security concerns of handling secrets.

Reply:
The reported "effective" config is what we choose to report. It is fine if we decide we want to mask certain keys. The purpose of the effective config is merely to allow OpAMP manager to show to the user what the current config is, it is for human consumption. Altering/masking keys will not break any other functionality, there is no expectation that the effective config is reported exactly as it is.

Reply:
I think we should leverage `configopaque.String` for this. I don't think the server should have access to the secrets on the configuration, if only because it makes thinking about the security implications much simpler.

Reply:
@MovieStoreGuy @mx-psi Right now the config that the extension receives is a raw, unfiltered `confmap.Conf` with the effective config. Do you have any pointers for how we could use `configopaque.String` to redact sensitive fields?

Reply:
The only way I can see to do this so far is to rebuild `otelcol.Config`, which doesn't feel right to me. The implementation would look something like this:

```golang
func (o *opampAgent) NotifyConfig(ctx context.Context, conf *confmap.Conf) error {
	receivers := conf.Get("receivers").(map[string]any)
	factories := otelcol.Factories{
		Receivers:  map[component.Type]receiver.Factory{},
		Extensions: map[component.Type]extension.Factory{},
		Exporters:  map[component.Type]exporter.Factory{},
	}
	for r, _ := range receivers {
		t := component.Type(r)
		f := o.host.GetFactory(component.KindReceiver, t)
		factories.Receivers[t] = f.(receiver.Factory)
	}

	// Provider that just returns `conf`
	mp := memoryProvider{conf: conf}
	p, _ := otelcol.NewConfigProvider(otelcol.ConfigProviderSettings{
		ResolverSettings: confmap.ResolverSettings{
			URIs: []string{"memory:otelcol"},
			Providers: map[string]confmap.Provider{
				mp.Scheme(): &mp,
			},
		},
	})
	cfg, err := p.Get(ctx, factories)

	redactedConf := confmap.New()
	redactedConf.Marshal(cfg)
	fmt.Printf("Conf: %+v\n", redactedConf.ToStringMap())
}
```

Reply:
It seems like ultimately we have to `reflect.TypeOf` on each field to check this, either through the Marshal/Unmarshal or more manually, since the way component developers express anything-config related is through the configuration structs.

Reply:
That was my suspicion, thanks for the confirmation. Unfortunately using reflection will very likely result in equal or more meddling with the config compared to recreating `otelcol.Config` since we will still need to convert the map object to component config structs. Re-implementing the config marshaling and unmarshaling process like this would be a fragile solution, I will try to solve this in the Collector APIs.